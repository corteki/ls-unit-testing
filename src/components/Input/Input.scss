.Input {
  position: relative;
  z-index: 1;
  display: flex;
  align-items: center;
  width: 100%;
  &__field {
    width: inherit;
    padding: 8px;
    margin: 0;
    &:focus {
      outline: none;
    }
  }
  &__clear {
    position: absolute;
    right: 32px;
    height: 20px;
    cursor: pointer;
  }
  &__open {
    position: absolute;
    right: 8px;
  }
}
