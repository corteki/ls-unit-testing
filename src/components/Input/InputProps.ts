import {MouseEvent, KeyboardEvent, ChangeEvent} from 'react';

export interface InputProps {
  value: string;
  placeholder: string;
  onClick: (e: MouseEvent<HTMLInputElement>) => void;
  onClear: (e: MouseEvent<HTMLInputElement>) => void;
  onInput: (e: ChangeEvent<HTMLInputElement>) => void;
}
