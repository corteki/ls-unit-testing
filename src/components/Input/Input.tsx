import React, { FC } from "react";
import { InputProps } from "./InputProps";
import ArrowDropDownIcon from '@material-ui/icons/ArrowDropDown';
import ClearIcon from '@material-ui/icons/Clear';
import "./Input.scss";

export const Input: FC<InputProps> = ({value, placeholder, onClick, onInput, onClear}) => (
  <div className="Input">
    <input
      className="Input__field"
      type="text"
      value={value}
      placeholder={placeholder}
      onClick={onClick}
      onChange={onInput}
    />
    {
      value.length > 0 &&
      <span 
        className="Input__clear"
        onClick={onClear}>
        <ClearIcon
          fontSize="small"/>
      </span>
    }
    <ArrowDropDownIcon 
      className="Input__open"
      fontSize="small"/>
  </div>
);
