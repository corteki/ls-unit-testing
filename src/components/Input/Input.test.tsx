import React from 'react';
import { render, screen, fireEvent } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { Input } from "./Input";

describe('<Input/>', () => {

  /**
   * [1] mocking event handlers
   */
  const MOCK_HANDLE_INPUT = jest.fn();
  const MOCK_HANDLE_CLEAR = jest.fn();
  const MOCK_HANDLE_CLICK = jest.fn();

  /**
   * [2] instance is available via "screen",
   * cleanup is done for us after each test
   */
  beforeEach(() => {
    render(
      <Input 
        value="MockValue" 
        placeholder="MockPlaceholder"
        onInput={MOCK_HANDLE_INPUT}
        onClear={MOCK_HANDLE_CLEAR}
        onClick={MOCK_HANDLE_CLICK}
      />
    );
  });

  /**
   * [3] verifying the default output of a input field
   */
  it('should render an empty input field without a clear icon and with a down arrow icon', () => {
    const tree = renderer
      .create(
        <Input 
          value="" 
          placeholder="MockPlaceholder"
          onInput={MOCK_HANDLE_INPUT}
          onClear={MOCK_HANDLE_CLEAR}
          onClick={MOCK_HANDLE_CLICK}
        />
      ).toJSON();
      expect(tree).toMatchSnapshot();
  });

  /**
   * [4] '@testing-library/react' allows to avoid testing implementation details
   */
  it('should render with a placeholder by default and render the value', () => {
    const input = screen.getByPlaceholderText(/MockPlaceholder/i);
    /**
     * [5] example of type specific matcher
     */
    expect(input).toHaveValue('MockValue');
  });

  it('should call the onInput handler when the input field is filled in', () => {
    const input = screen.getByPlaceholderText(/MockPlaceholder/i);
    const mockEvent = { target: { value: 'MockInput' } };

    fireEvent.focus(input);
    fireEvent.change(input, mockEvent);

    expect(MOCK_HANDLE_INPUT).toHaveBeenCalled();
  });

  it('should call the onClear handler when the clear icon is clicked', () => {
    const clearIcon = screen.getByTestId('clear');

    fireEvent.click(clearIcon);

    expect(MOCK_HANDLE_CLEAR).toHaveBeenCalled();
  });

  it('should call the onClick handler when the input field is clicked', () => {
    const input = screen.getByPlaceholderText(/MockPlaceholder/i);

    fireEvent.click(input);

    expect(MOCK_HANDLE_CLICK).toHaveBeenCalled();
  });

});
