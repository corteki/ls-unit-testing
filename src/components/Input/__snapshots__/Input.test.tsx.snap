// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Input/> should render an empty input field without a clear icon and with a down arrow icon 1`] = `
<div
  className="Input"
>
  <input
    className="Input__field"
    onChange={[MockFunction]}
    onClick={[MockFunction]}
    placeholder="MockPlaceholder"
    type="text"
    value=""
  />
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root Input__open MuiSvgIcon-fontSizeSmall"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
  </svg>
</div>
`;

exports[`<Input/> snapshots should render an empty input field without a clear icon and with a down arrow icon 1`] = `
<div
  className="Input"
>
  <input
    className="Input__field"
    onChange={[MockFunction]}
    onClick={[MockFunction]}
    placeholder="MockPlaceholder"
    type="text"
    value=""
  />
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root Input__open MuiSvgIcon-fontSizeSmall"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
  </svg>
</div>
`;

exports[`<Input/> snapshots should render an input field with a clear icon and with a down arrow icon 1`] = `
<div
  className="Input"
>
  <input
    className="Input__field"
    onChange={[MockFunction]}
    onClick={[MockFunction]}
    placeholder="MockPlaceholder"
    type="text"
    value="myValue"
  />
  <span
    className="Input__clear"
    data-testid="clear"
    onClick={[MockFunction]}
  >
    <svg
      aria-hidden={true}
      className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall"
      focusable="false"
      viewBox="0 0 24 24"
    >
      <path
        d="M19 6.41L17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"
      />
    </svg>
  </span>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root Input__open MuiSvgIcon-fontSizeSmall"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
  </svg>
</div>
`;
